libdnf/sack/query.cpp

```diff
@@ -0,0 +1,1821 @@
+                (pkg_epoch == epoch && cmp_type & HY_EQ))
+                MAPSET(m, id);
+        }
+    }
+}
+
+void
+Query::Impl::filterEvr(const Filter & f, Map *m)
+{
+    Pool *pool = dnf_sack_get_pool(sack);
+    int cmp_type = f.getCmpType();
+    auto resultPset = result.get();
+
+    for (auto match : f.getMatches()) {
+        Id match_evr = pool_str2id(pool, match.str, 1);
+
+        Id id = -1;
+        while (true) {
+            id = resultPset->next(id);
+            if (id == -1)
+                break;
+            Solvable *s = pool_id2solvable(pool, id);
+            int cmp = pool_evrcmp(pool, s->evr, match_evr, EVRCMP_COMPARE);
+
+            if ((cmp > 0 && cmp_type & HY_GT) || (cmp < 0 && cmp_type & HY_LT) ||
+                (cmp == 0 && cmp_type & HY_EQ)) {
+                MAPSET(m, id);
+            }
+        }
+    }
+}
+
+void
+Query::Impl::filterNevra(const Filter & f, Map *m)
+{
+    Pool *pool = dnf_sack_get_pool(sack);
+    int cmp_type = f.getCmpType();
+    int fn_flags = (HY_ICASE & cmp_type) ? FNM_CASEFOLD : 0;
+    auto resultPset = result.get();
+
+    for (auto match : f.getMatches()) {
+        const char *nevra_pattern = match.str;
+        if (strpbrk(nevra_pattern, "(/=<> "))
+            continue;
+
+        gboolean present_epoch = strchr(nevra_pattern, ':') != NULL;
+
+        Id id = -1;
+        while (true) {
+            id = resultPset->next(id);
+            if (id == -1)
+                break;
+            Solvable* s = pool_id2solvable(pool, id);
+
+            char* nevra = pool_solvable_epoch_optional_2str(pool, s, present_epoch);
+            if (!(HY_GLOB & cmp_type)) {
+                if (HY_ICASE & cmp_type) {
+                    if (strcasecmp(nevra_pattern, nevra) == 0)
+                        MAPSET(m, id);
+                } else {
+                    if (strcmp(nevra_pattern, nevra) == 0)
+                        MAPSET(m, id);
+                }
+            } else if (fnmatch(nevra_pattern, nevra, fn_flags) == 0) {
+                MAPSET(m, id);
+            }
+        }
+    }
+}
+
+void
+Query::Impl::filterVersion(const Filter & f, Map *m)
+{
+    Pool *pool = dnf_sack_get_pool(sack);
+    int cmp_type = f.getCmpType();
+    auto resultPset = result.get();
+
+    for (auto match_in : f.getMatches()) {
+        const char *match = match_in.str;
+        char *filter_vr = solv_dupjoin(match, "-0", NULL);
+
+        Id id = -1;
+        while (true) {
+            id = resultPset->next(id);
+            if (id == -1)
+                break;
+            char *e, *v, *r;
+            Solvable *s = pool_id2solvable(pool, id);
+            if (s->evr == ID_EMPTY)
+                continue;
+            const char *evr = pool_id2str(pool, s->evr);
+
+            pool_split_evr(pool, evr, &e, &v, &r);
+
+            if (cmp_type == HY_GLOB) {
+                if (fnmatch(match, v, 0))
+                    continue;
+                MAPSET(m, id);
```

Comment:
We probably want to continue next iteration after this line. Another version of code.
```
            if (cmp_type == HY_GLOB) {
                if (fnmatch(match, v, 0) == 0)
                    MAPSET(m, id);
            } else {
                char *vr = pool_tmpjoin(pool, v, "-0", NULL);
                int cmp = pool_evrcmp_str(pool, vr, filter_vr, EVRCMP_COMPARE);
                if ((cmp > 0 && cmp_type & HY_GT) ||
                    (cmp < 0 && cmp_type & HY_LT) ||
                    (cmp == 0 && cmp_type & HY_EQ)) {
                    MAPSET(m, id);
                }
            }
```